### PR TITLE
DualStack config Procedure

### PR DESCRIPTION
**Pull request by jmanthei**
Version(s):

Issue: https://issues.redhat.com/browse/OSDOCS-4938

Link to docs preview: https://66173--docspreview.netlify.app/openshift-enterprise/latest/networking/hardware_networks/configuring-sriov-net-attach#nw-multus-configure-dualstack-ip-address_configuring-sriov-net-attach

QE review:
- [ ] QE has approved this change.

Additional information:

### CODE REVIEW

File: modules/nw-multus-ipam-object.adoc [hunk 1]
```diff
@@ -310,4 +310,47 @@ pod/whereabouts-reconciler-p4sxw 1/1 Running 0 6s
 pod/whereabouts-reconciler-rvfdv 1/1 Running 0 6s
 pod/whereabouts-reconciler-svzw9 1/1 Running 0 6s
 daemonset.apps/whereabouts-reconciler 6 6 6 6 6 kubernetes.io/os=linux 6s
-----
\ No newline at end of file
+----
+
+[id="nw-multus-creating-whereabouts-dualstack-config_{context}"]
+== Creating a DualStack config
```
Comment (dougbtv) [inline modules/nw-multus-ipam-object.adoc:316]: nit: I think this can be "Dual stack" or "dual-stack" (unless we have a precedent set for DualStack in other docs)

File: modules/nw-multus-ipam-object.adoc [hunk 2]
```diff
@@ -310,4 +310,47 @@ pod/whereabouts-reconciler-p4sxw 1/1 Running 0 6s
...
+
+DualStack can be configured with `ipRanges`.
```
Comment (dougbtv) [inline modules/nw-multus-ipam-object.adoc:318]: suggestion: 

> Dual stack IP address assignment (for both IPv4 and IPv6 addresses), or multiple IP address assignment, can be configured with the `ipRanges` parameter.

I'm trying to get across that dual stack is the primary use case, but, you can configure n-number of addresses, and I want folks to know that "multiple IP address assignment" works too.

File: modules/nw-multus-ipam-object.adoc [hunk 3]
```diff
@@ -310,4 +310,47 @@ pod/whereabouts-reconciler-p4sxw 1/1 Running 0 6s
...
+         }, {
+           "range": "abcd: :1/64"
+         }]
+       }
+      }
+
+
+----
+
+. INSERT A CHECK STEP HERE SO WE KNOW WE DID IT CORRECTLY; THIS SHOULD BE A VERIFICATION STEP.
```
Comment (dougbtv) [inline modules/nw-multus-ipam-object.adoc:356]: Example of how we attach this network to a pod: https://docs.openshift.com/container-platform/4.13/networking/multiple_networks/attaching-pod.html#nw-multus-add-pod_attaching-pod

Then, to verify that you have all ip addresses assigned, you can exec:

```
oc exec -it mypod -- ip a
```

Caveat: Required that the `ip` binary is available in the pod.

and also

```
oc describe pod mypod
```

And look for the annotation called `k8s.v1.cni.cncf.io/network-status` (see step 3 in the above docs link) which should show both IP addresses assigned as meta-data associated with the pod.

File: modules/nw-multus-ipam-object.adoc [hunk 4]
```diff
@@ -310,4 +310,47 @@ pod/whereabouts-reconciler-p4sxw 1/1 Running 0 6s
...
+      {
+       "name": "whereabouts-shim",
+       "cniVersion": "0.3.1,
+       "type": "bridge",
+       "ipam": {
+         "type": "whereabouts",
+         "ipRanges": [{
+           "range": "192.168.10.1/24"
+         }, {
+           "range": "abcd: :1/64"
```
Comment (dougbtv) [inline modules/nw-multus-ipam-object.adoc:348]: let's use a documentation range for IPv6, from: https://www.rfc-editor.org/rfc/rfc3849

Such as: 

```
2001:db8::/64
```

File: modules/nw-multus-ipam-object.adoc (hunk 1 above)
Comment (dougbtv) [reply modules/nw-multus-ipam-object.adoc:316]: Also maybe we want to specifically say "Creating a configuration for assignment of dual-stack IP addresses dynamically" which is more specific to what Whereabouts actually does here. (maybe too wordy, but, as a start)

File: modules/nw-multus-ipam-object.adoc [hunk 5]
```diff
@@ -310,4 +310,47 @@ pod/whereabouts-reconciler-p4sxw 1/1 Running 0 6s
...
+kind: Network
+metadata:
+  name: cluster
+spec:
+  additionalNetworks:
+  - name: whereabouts-shim
+    namespace: default
+    rawCNIConfig: |-
+      {
+       "name": "whereabouts-shim",
```
Comment (dougbtv) [inline modules/nw-multus-ipam-object.adoc:340]: While the name is arbitrary, let's use another name here for purposes of the example, and since the shim is kind of a specific thing, maybe:

```
"name": "whereabouts-dual-stack",
```

**Review by dougbtv — CHANGES_REQUESTED** [review]

File: modules/nw-multus-ipam-object.adoc (hunk 1 above)
Comment (jmanthei) [reply modules/nw-multus-ipam-object.adoc:316]: Dualstack is frequently used as one word in docs but I will absolutely double-check to ensure this is the right context!

File: modules/nw-multus-ipam-object.adoc (hunk 5 above)
Comment (jmanthei) [reply modules/nw-multus-ipam-object.adoc:340]: Thank you!

File: modules/nw-multus-ipam-object.adoc [hunk 6]
```diff
@@ -310,4 +310,62 @@ pod/whereabouts-reconciler-p4sxw 1/1 Running 0 6s
 pod/whereabouts-reconciler-rvfdv 1/1 Running 0 6s
 pod/whereabouts-reconciler-svzw9 1/1 Running 0 6s
 daemonset.apps/whereabouts-reconciler 6 6 6 6 6 kubernetes.io/os=linux 6s
+----
+
+[id="nw-multus-creating-whereabouts-dualstack-config_{context}"]
+== Creating a configuration for assignment of dualstack IP addresses dynamically
+
+DualStack IP address assignment can be configured with the `ipRanges` parameter for:
```
Comment (dougbtv) [inline modules/nw-multus-ipam-object.adoc:318]: nit: Let's change to "dual-stack" with a hyphen. On this line and the one just above it. Looks like it's the most common in the docs. Here's my count of the variations in the docs:

```
$ grep -Prin "dualstack" * | wc -l
20
$ grep -Prin "dual.stack" * | wc -l
124
$ grep -Prin "dual stack" * | wc -l
24
$ grep -Prin "dual\-stack" * | wc -l
100
```

File: modules/nw-multus-ipam-object.adoc [hunk 7]
```diff
@@ -310,4 +310,62 @@ pod/whereabouts-reconciler-p4sxw 1/1 Running 0 6s
...
+- IPv4 addresses
+- IPv6 addresses
+- multiple IP address assignment 
+
+.Procedure
+
+. Set type to whereabouts
+. Use `ipRanges` to allocate IP addresses.
+
+For example, defining the range as `192.168.10.1/24` excludes the first network address and the last broadcast address.
```
Comment (dougbtv) [inline modules/nw-multus-ipam-object.adoc:328]: I think we can omit this line? I think we cover setting a CIDR range elsewhere in the Whereabouts docs.

File: modules/nw-multus-ipam-object.adoc [hunk 8]
```diff
@@ -310,4 +310,62 @@ pod/whereabouts-reconciler-p4sxw 1/1 Running 0 6s
...
+
+[source,yaml]
+----
+cniVersion: operator.openshift.io/v1
+kind: Network
+metadata:
+  name: cluster
+spec:
+  additionalNetworks:
+  - name: whereabouts-shim
```
Comment (dougbtv) [inline modules/nw-multus-ipam-object.adoc:339]: Let's change this to `- name: whereabouts-dual-stack`

File: modules/nw-multus-ipam-object.adoc [hunk 9]
```diff
@@ -310,4 +310,62 @@ pod/whereabouts-reconciler-p4sxw 1/1 Running 0 6s
...
+         }]
+       }
+      }
+
+
+----
+
+. Verify that all IP addresses are assigned
+
+Run the following two commands to ensure the IP adresses are assigned as meta-data.
```
Comment (dougbtv) [inline modules/nw-multus-ipam-object.adoc:361]: With these kind of configurations, there's kind of two steps -- defining a network configuration, and then associating it with a pod. So the user won't see anything until they also associate it with a pod.

So maybe we can link to: https://docs.openshift.com/container-platform/4.13/networking/multiple_networks/attaching-pod.html#nw-multus-add-pod_attaching-pod which shows how to attach a network. And then change it to like:

> After [attaching this additional network to a pod], verify the IP address assignment by running these two commands to ensure the IP addresses are assigned, and is present in the pod metadata in the  `k8s.v1.cni.cncf.io/network-status` annotation:

(where the square brackets indicates a link pointed to the above location in docs -- and of course, modified to fit your style guide)

**Review by dougbtv — APPROVED** [review]

**Review by dougbtv — CHANGES_REQUESTED** [review]
(oops I reviewed with an approve, meant to request changes)

File: modules/nw-multus-ipam-object.adoc (hunk 3 above)
Comment (maiqueb) [reply modules/nw-multus-ipam-object.adoc]: FWIW, and assuming we have `jq` available (not required, but gives a better formatted output) we can:
```bash
oc get pod mypod -ojsonpath="{ @.metadata.annotations.k8s\.v1\.cni\.cncf\.io\/network-status }" | jq
```

This is pretty much what Doug is saying above, but instead of "looking for", you ask for this particular annotation value. 

note: the `| jq` at the end is optional.

File: modules/nw-multus-ipam-object.adoc (hunk 6 above)
Comment (jmanthei) [reply modules/nw-multus-ipam-object.adoc:318]: Sure thing

File: modules/nw-multus-ipam-object.adoc (hunk 8 above)
Comment (jmanthei) [reply modules/nw-multus-ipam-object.adoc:339]: done

File: modules/nw-multus-ipam-object.adoc (hunk 7 above)
Comment (jmanthei) [reply modules/nw-multus-ipam-object.adoc:328]: It seemed like it may not have been totally relevant but I leave that to your discretion

**Comment by jmanthei** [comment]
@maiqueb QE check please

**Comment by jmanthei** [comment]
...I broke the build preview again trying to restore one thing...code should be correct and won't change; just the xref issue.

**Comment by kalexand-rh** [comment]
The build is passing now. @maiqueb, please review this change.

File: modules/nw-multus-configure-dualstack-ip-address.adoc [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+
+. Attach network to a pod. For more information, see "Adding a pod to an additional network".
+
+. Verify that all IP addresses are assigned.
+
+Run the following two commands to ensure the IP adresses are assigned as meta-data.
+
+[source,yaml]
+----
+oc exec -it mypod -- ip a
```
Comment (maiqueb) [inline modules/nw-multus-configure-dualstack-ip-address.adoc:56]: As @dougbtv said, the IP command might not be available in the container image. I would only present the other alternative, which is always available.

File: modules/nw-multus-configure-dualstack-ip-address.adoc [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+----
+
+[source,yaml]
+----
+oc describe pod mypod
```
Comment (maiqueb) [inline modules/nw-multus-configure-dualstack-ip-address.adoc:61]: I would either get the specific network-status by specifying jsonpath - i.e. `oc get -ojsonpath ...`  or put some grep to directly show the interesting data.

**Review by maiqueb — COMMENTED** [review]
Some minor comments.

**Comment by jmanthei** [comment]
Closed here to https://github.com/openshift/openshift-docs/pull/66646

File: modules/nw-multus-configure-dualstack-ip-address.adoc (hunk 10 above)
Comment (jmanthei) [reply modules/nw-multus-configure-dualstack-ip-address.adoc:56]: Ok

File: modules/nw-multus-configure-dualstack-ip-address.adoc [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+       "name": "whereabouts-dual-stack",
+       "cniVersion": "0.3.1,
+       "type": "bridge",
+       "ipam": {
+         "type": "whereabouts",
+         "ipRanges": [{
+           "range": "192.168.10.1/24"
+         }, {
+           "range": "2001:db8::/64"
+         }]
```
[the remark this answers is not in the document]
Comment (jmanthei) [reply modules/nw-multus-configure-dualstack-ip-address.adoc:41]: done

File: modules/nw-multus-configure-dualstack-ip-address.adoc (hunk 11 above)
Comment (jmanthei) [reply modules/nw-multus-configure-dualstack-ip-address.adoc:61]: If you want a different code set, I will need that directly from you.

Comment (maiqueb) [reply modules/nw-multus-configure-dualstack-ip-address.adoc:61]: I suggested it in https://github.com/openshift/openshift-docs/pull/66173#discussion_r1360362398

Let me know if that makes sense to you. I think this is more straight-forward to the user, but it might raise some questions (having to escape some characters, etc).

Also, let's get @dougbtv 's opinion.
